### PR TITLE
cpu/lpc2387: PM: enable SLEEP & POWERDOWN mode

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

I'm not sure what it was that this wasn't working before, but when re-initializing PLL (& Flash) the sleep modes on this MCU work just fine.

### Testing procedure

Run `tests/periph_pm`.

 - `pm set 2`: wake up by button or RTC (`set_rtc 2`)
 - `pm set 1`: wake up by button or RTC (`set_rtc 1`)
 - `pm set 0`: will not wake by button, only RTC (`set_rtc 0`) - CPU will reset

### Issues/PRs references

Requires #13757 for `set_rtc`

### CODE REVIEW

**Comment by leandrolanzieri** [comment]
I tried this on the `msba2` with the RTC command, but it does not seem to come back from the sleep:
```
2020-04-29 19:08:11,879 #main(): This is RIOT! (Version: 2020.07-devel-303-ge1f2f-cpu/lpc2387/pm_fix)
2020-04-29 19:08:11,881 # This application allows you to test the CPU power management.
2020-04-29 19:08:11,884 # The available power modes are 0 - 2. Lower-numbered power modes
2020-04-29 19:08:11,887 # save more power, but may require an event/interrupt to wake up
2020-04-29 19:08:11,889 # the CPU. Reset the CPU if needed.
2020-04-29 19:08:11,891 # mode 0 blockers: 1 
2020-04-29 19:08:11,892 # mode 1 blockers: 1 
2020-04-29 19:08:11,892 # mode 2 blockers: 1 
2020-04-29 19:08:11,894 # Lowest allowed mode: 3
set_rtc 2 1

# Here I pressed reset
2020-04-29 19:08:16,159 #  set_rtc 2 1
2020-04-29 19:08:26,290 # Setting power mode 2 for 1 seconds�main(): This is RIOT! (Version: 2020.07-devel-303-ge1f2f-cpu/lpc2387/pm_fix)
```

**Comment by benpicco** [comment]
It comes back, but the UART output doesn't yet work in the ISR handler because that gets called before `cpu_init_pll()` - shell should work again after that.

The waking interrupt should probably call some `cpu_woke_from_sleep()` function - something like that would be needed on other MCUs too. (e.g. sam0 has `cpu_pm_cb_leave()` which gets called after `cortexm_sleep()` returns - but the wake-up ISR will be called before that.

**Comment by maribu** [comment]
For me it works: I can wake the MSB-A2 up from both pm mode 1 and 2 using both the RTC and via "button" (I configured P2.2 as BTN0 with internal pull up and used a jumper wire to "press" the button).

I can confirm the stdio issue during the ISR directly after the wake up. But using stdio from IRQ context is IMO something that comes with "use at your own risk" label and is only really justifiable for `DEBUG()`.

**Comment by leandrolanzieri** [comment]
I tried using the pin interrupt as source of wake up and then it works. I get the message and the shell works again. But with the RTC it still does not come back, not even the shell :confused:

**Comment by maribu** [comment]
I might have an idea: I set the current time in the RTC first. Maybe there is an issue when the year in the RTC is uninitialized to a random value?

**Comment by maribu** [comment]
I can confirm that the RTC alarm does not work when the year is above 4753. But that issues is unrelated to power management.

So: Power management works as expected :-)

File: cpu/lpc2387/clocks.c [hunk 1]
```diff
@@ -43,14 +43,14 @@ static inline void pllfeed(void)
...
     MAMCR  = 0x0000;
     MAMTIM = 0x0003;
     MAMCR  = 0x0002;
 }
 
-static void init_clks1(void)
+void cpu_init_pll(void)
 {
     /* Disconnect PLL */
     PLLCON &= ~0x0002;
```
Comment (maribu) [inline cpu/lpc2387/clocks.c:56]: Care to also change `while (PLLSTAT & BIT25);` to `while (PLLSTAT & BIT25) {}` in line 59? And same style fix in line 65, line 70, line 93, and line 100?

File: cpu/lpc2387/periph/pm.c [hunk 2]
```diff
@@ -27,25 +27,36 @@
 #define ENABLE_DEBUG (0)
 #include "debug.h"
 
+extern void cpu_init_pll(void);
+extern void cpu_init_mam(void);
```
Comment (maribu) [inline cpu/lpc2387/periph/pm.c:31]: Why not just add those to `cpu/lpc2387/include/lpc2387.h`?

**Comment by benpicco** [comment]
> I can confirm that the RTC alarm does not work when the year is above 4753. But that issues is unrelated to power management.

Uh that's due to that power-on reset not being detected properly on the MSB-A2, right?
The easiest way to fix this would probably be to add a `RESET_RTC_ON_INIT` to `board.h`

**Comment by maribu** [comment]
But shouldn't the RTC still be usable in 2750 years from now? ;-P

**Comment by benpicco** [comment]
Or how about

```patch
--- a/cpu/lpc2387/periph/rtc.c
+++ b/cpu/lpc2387/periph/rtc.c
@@ -56,7 +56,8 @@ void rtc_init(void)
     /* Initialize clock to a a sane and predictable default
      * after cold boot or external reset.
      */
-    if ((RSIR == RSIR_POR) || (RSIR == (RSIR_POR | RSIR_EXTR))) {
+    if ((RSIR == RSIR_POR) || (RSIR == (RSIR_POR | RSIR_EXTR))
+       || RTC_YEAR > 250) {
         struct tm localt = { .tm_year = 70 };
         rtc_set_time(&localt);
     }
```

Assuming no `msba2`s will be in use anymore after the year 2150 :wink:

**Comment by maribu** [comment]
> Assuming no `msba2`s will be in use anymore after the year 2150 😉

For the MSB-A2s I have on my desktop, this would work pretty reliable. Apparently the PUF of the SRAM storing the RTC on my device has a tendency to be in a year > 4000 when uninitialized. But other people could be less lucky. I think the `RESET_RTC_ON_INIT` is the better option here, as it is predictable. If I don't forget, I create a PR this weekend.

**Comment by benpicco** [comment]
> Apparently the PUF of the SRAM storing the RTC on my device has a tendency to be in a year > 4000 when uninitialized. But other people could be less lucky.

Well the year counter is 12 bit, so there is a ~94% chance it will be at a value > 250 if it's entirely random. 

> I think the `RESET_RTC_ON_INIT` is the better option here, as it is predictable. If I don't forget, I create a PR this weekend.

The downside of that is that it would lose time across reboots or deep sleep then

**Comment by maribu** [comment]
> Well the year counter is 12 bit, so there is a ~94% chance it will be at a value > 250 if it's entirely random.

I could swear that I get the same year when uninitialized for my MSB-A2 every time, but I didn't write it down to compare. I guess if someone is unlucky to have a MSB-A2 with a tendency to have the RTC value below 250, that person will end up with an uninitialized value most of the time.

> The downside of that is that it would lose time across reboots or deep sleep then

The MSB-A2 are only legacy systems of a former testbed. I guess if they ever get uses regularly again, than it will be a similar use case. In that it is more important to have predictable RTC values, than having correct values.

**Pull request merged**